utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,458 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_char16trie::char16trie::{Char16Trie, TrieResult};
+    /// use zerovec::ZeroVec;
+    ///
+    /// // A Char16Trie containing the ASCII characters 'a' and 'b'.
+    /// let trie_data = vec![48, 97, 176, 98, 32868];
+    /// let trie = Char16Trie::new(ZeroVec::from_slice(trie_data.as_slice()));
+    ///
+    /// let mut iter = trie.iter();
+    /// let res = iter.next('a');
+    /// assert_eq!(res, TrieResult::Intermediate(1));
+    /// let res = iter.next('b');
+    /// assert_eq!(res, TrieResult::FinalValue(100));
+    /// let res = iter.next('c');
+    /// assert_eq!(res, TrieResult::NoMatch);
+    /// ```
+    pub fn next_u32(&mut self, c: u32) -> TrieResult {
+        if c <= 0xffff {
+            self.next_u16(c as u16)
+        } else {
+            match self.next_u16(u16_lead(c as i32)) {
+                TrieResult::NoValue | TrieResult::Intermediate(_) => {
+                    self.next_u16(u16_tail(c as i32))
+                }
+                _ => TrieResult::NoMatch,
+            }
+        }
+    }
+
+    /// Traverses the trie from the current state for this input char.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_char16trie::char16trie::{Char16Trie, TrieResult};
+    /// use zerovec::ZeroVec;
+    ///
+    /// // A Char16Trie containing the ASCII characters 'a' and 'b'.
+    /// let trie_data = vec![48, 97, 176, 98, 32868];
+    /// let trie = Char16Trie::new(ZeroVec::from_slice(trie_data.as_slice()));
+    ///
+    /// let mut iter = trie.iter();
+    /// let res = iter.next_u16('a' as u16);
+    /// assert_eq!(res, TrieResult::Intermediate(1));
+    /// let res = iter.next_u16('b' as u16);
+    /// assert_eq!(res, TrieResult::FinalValue(100));
+    /// let res = iter.next_u16('c' as u16);
+    /// assert_eq!(res, TrieResult::NoMatch);
+    /// ```
+    pub fn next_u16(&mut self, c: u16) -> TrieResult {
+        if self.pos.is_none() {
+            return TrieResult::NoMatch;
+        }
+        let mut pos = self.pos.unwrap();
+        if let Some(length) = self.remaining_match_length {
+            // Remaining part of a linear-match node
+            if c == self.get(pos) {
+                pos += 1;
+                self.pos = Some(pos);
+                if length == 0 {
+                    self.remaining_match_length = None;
+                    let node = self.get(pos);
+                    if node >= MIN_VALUE_LEAD {
+                        return self.value_result(pos);
+                    }
+                } else {
+                    self.remaining_match_length = Some(length - 1);
+                }
+                return TrieResult::NoValue;
+            }
+            self.stop();
+            TrieResult::NoMatch
+        } else {
+            self.next_impl(pos, c)
+        }
+    }
+
+    fn get(&self, pos: usize) -> u16 {
+        u16::from_unaligned(self.trie[pos])
```

Comment:
Issue: this is panicky.

Another occurrence of #1183

Reply:
My plan is to land this in experimental. I've filed https://github.com/unicode-org/icu4x/issues/1440 to follow up on fixing this.